rclcpp_action/src/server.cpp

```diff
@@ -12,30 +12,477 @@
+    wait_set,
+    pimpl_->action_server_.get(),
+    &pimpl_->goal_request_ready_,
+    &pimpl_->cancel_request_ready_,
+    &pimpl_->result_request_ready_,
+    &pimpl_->goal_expired_);
+
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  return pimpl_->goal_request_ready_ ||
+         pimpl_->cancel_request_ready_ ||
+         pimpl_->result_request_ready_;
+}
+
+void
+ServerBase::execute()
+{
+  if (pimpl_->goal_request_ready_) {
+    execute_goal_request_received();
+  } else if (pimpl_->cancel_request_ready_) {
+    execute_cancel_request_received();
+  } else if (pimpl_->result_request_ready_) {
+    execute_result_request_received();
+  } else if (pimpl_->goal_expired_) {
+    execute_check_expired_goals();
+  } else {
+    throw std::runtime_error("Executing action server but nothing is ready");
+  }
+}
+
+void
+ServerBase::execute_goal_request_received()
+{
+  rcl_ret_t ret;
+  rcl_action_goal_info_t goal_info = rcl_action_get_zero_initialized_goal_info();
+  rmw_request_id_t request_header;
+
+  std::shared_ptr<void> message = create_goal_request();
+  ret = rcl_action_take_goal_request(
+    pimpl_->action_server_.get(),
+    &request_header,
+    message.get());
+
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  pimpl_->goal_request_ready_ = false;
+  std::array<uint8_t, 16> uuid = get_goal_id_from_goal_request(message.get());
+  for (size_t i = 0; i < 16; ++i) {
+    goal_info.uuid[i] = uuid[i];
+  }
+
+  // Call user's callback, getting the user's response and a ros message to send back
+  auto response_pair = call_handle_goal_callback(uuid, message);
+
+  ret = rcl_action_send_goal_response(
+    pimpl_->action_server_.get(),
+    &request_header,
+    response_pair.second.get());
+
+  if (RCL_RET_OK != ret) {
+    rclcpp::exceptions::throw_from_rcl_error(ret);
+  }
+
+  // if goal is accepted, create a goal handle, and store it
+  if (GoalResponse::ACCEPT == response_pair.first) {
+    // rcl_action will set time stamp
+    auto deleter = [](rcl_action_goal_handle_t * ptr)
+      {
+        if (nullptr != ptr) {
+          rcl_ret_t fail_ret = rcl_action_goal_handle_fini(ptr);
+          (void)fail_ret;  // TODO(sloretz) do something with error during cleanup
+          delete ptr;
+        }
+      };
+    rcl_action_goal_handle_t * rcl_handle = rcl_action_accept_new_goal(
+      pimpl_->action_server_.get(), &goal_info);
+    if (!rcl_handle) {
+      throw std::runtime_error("Failed to accept new goal\n");
+    }
+
+    std::shared_ptr<rcl_action_goal_handle_t> handle(new rcl_action_goal_handle_t, deleter);
+    // Copy out goal handle since action server storage disappears when it is fini'd
+    *handle = *rcl_handle;
+
+    // publish status since a goal's state has changed (was accepted)
+    publish_status();
+
+    // Change status to executing
+    ret = rcl_action_update_goal_state(handle.get(), GOAL_EVENT_EXECUTE);
```

Comment:
One problem with this approach is an action server that wants to execute goals sequentially while keeping a queue of goals will report all of them as executing in the status message.

One way to fix this would be to add a `GoalResponse` called `ACCEPT_AND_DEFER` that skipped the automatic transition to executing. The user would become responsible for calling a method on the goal handle to begin execution. This would need another callback between the goal handle and `ServerBase` that could trigger publishing another status message.

Reply:
Why do we have to transition to `EXECUTING` right away? Can't we stay in `ACCEPTED`?

Reply:
You're write it can stay in accepted, just need to add the plumbing from the goal handle to allow the user to transition to executing on their own time.

Reply:
Actually, if the user accepts it they won't have a goal handle. The first time it's sent to the user is via the `handle_execute` callback. Need to rethink this one

Reply:
After talking offline I changed `handle_execute` to `handle_accepted` in 339a36e85039c6d95dbd8e553c29f311f86dd0ed

Reply:
Added all the stuff I missed and a test for deferred executing in e7fd566c1c38d9d0ac4710db3cda1632efca4526